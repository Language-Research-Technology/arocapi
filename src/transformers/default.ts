import type { Entity, File } from '../generated/prisma/client.js';

/**
 * Standard entity shape - output of base transformation
 * Does not include access information
 */
export type StandardEntity = {
  id: string;
  name: string;
  description: string;
  memberOf: string | null;
  rootCollection: string | null;
  metadataLicenseId: string;
  contentLicenseId: string;
} & ({ entityType: string; fileId?: never } | { entityType: 'http://schema.org/MediaObject'; fileId: string });

/**
 * Access information for an entity
 */
type AccessInfo = {
  metadata: boolean;
  content: boolean;
  contentAuthorizationUrl?: string;
};

/**
 * Authorised entity - includes access information
 * This is the output of the access transformer
 */
export type AuthorisedEntity = StandardEntity & {
  access: AccessInfo;
};

/**
 * Access information for a file
 * Files only control content access - metadata is always accessible
 */
type FileAccessInfo = {
  content: boolean;
  contentAuthorizationUrl?: string;
};

/**
 * Standard file shape - output of base file transformation
 * Does not include access information
 * Files only have contentLicenseId (no metadataLicenseId)
 */
export type StandardFile = {
  id: string;
  filename: string;
  mediaType: string;
  size: number;
  memberOf: string;
  rootCollection: string;
  contentLicenseId: string;
};

/**
 * Authorised file - includes access information
 * This is the output of the file access transformer
 * File metadata is always accessible - only content access is controlled
 */
export type AuthorisedFile = StandardFile & {
  access: FileAccessInfo;
};

/**
 * Base entity transformer - always applied first
 * Transforms raw database entity to standard entity shape (without access)
 */
export const baseEntityTransformer = (entity: Entity): StandardEntity => {
  const base: StandardEntity = {
    id: entity.rocrateId,
    name: entity.name,
    description: entity.description,
    entityType: entity.entityType,
    memberOf: entity.memberOf,
    rootCollection: entity.rootCollection,
    metadataLicenseId: entity.metadataLicenseId,
    contentLicenseId: entity.contentLicenseId,
  };

  if (base.entityType === ('http://schema.org/MediaObject' as const)) {
    if (!entity.fileId) {
      return base;
    }

    return {
      ...base,
      entityType: base.entityType,
      fileId: entity.fileId,
    };
  }

  return base;
};

/**
 * All Public Access Transformer - grants full access to metadata and content
 *
 * WARNING: This transformer makes ALL content publicly accessible without restrictions.
 * Only use this for fully public datasets where no access control is needed.
 *
 * For repositories with restricted content, implement a custom accessTransformer
 * that checks user permissions and licenses.
 *
 * @example
 * ```typescript
 * await server.register(arocapi, {
 *   prisma,
 *   opensearch,
 *   accessTransformer: AllPublicAccessTransformer, // Explicit choice for public data
 * });
 * ```
 */
export const AllPublicAccessTransformer = (entity: StandardEntity): AuthorisedEntity => ({
  ...entity,
  access: {
    metadata: true,
    content: true,
  },
});

/**
 * Base file transformer - always applied first
 * Transforms raw database file to standard file shape (without access)
 */
export const baseFileTransformer = (file: File): StandardFile => ({
  id: file.fileId,
  filename: file.filename,
  mediaType: file.mediaType,
  size: Number(file.size),
  memberOf: file.memberOf,
  rootCollection: file.rootCollection,
  contentLicenseId: file.contentLicenseId,
});

/**
 * All Public File Access Transformer - grants full access to file content
 *
 * WARNING: This transformer makes ALL file content publicly accessible without restrictions.
 * Only use this for fully public datasets where no access control is needed.
 *
 * For repositories with restricted content, implement a custom fileAccessTransformer
 * that checks user permissions and licenses.
 *
 * Note: File metadata (filename, size, mediaType, etc.) is always accessible.
 * This transformer only controls content access.
 *
 * @example
 * ```typescript
 * await server.register(arocapi, {
 *   prisma,
 *   opensearch,
 *   accessTransformer: AllPublicAccessTransformer,
 *   fileAccessTransformer: AllPublicFileAccessTransformer, // Explicit choice for public data
 * });
 * ```
 */
export const AllPublicFileAccessTransformer = (file: StandardFile): AuthorisedFile => ({
  ...file,
  access: {
    content: true,
  },
});
