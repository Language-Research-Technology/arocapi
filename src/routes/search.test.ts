import { afterEach, beforeEach, describe, expect, it } from 'vitest';

import { fastify, fastifyAfter, fastifyBefore, opensearch, prisma } from '../test/helpers/fastify.js';
import { AllPublicAccessTransformer } from '../transformers/default.js';
import searchRoute from './search.js';

describe('Search Route', () => {
  beforeEach(async () => {
    await fastifyBefore();
    await fastify.register(searchRoute, { accessTransformer: AllPublicAccessTransformer });
  });

  afterEach(async () => {
    await fastifyAfter();
  });

  describe('POST /search', () => {
    it('should perform basic search successfully', async () => {
      const mockEntities = [
        {
          rocrateId: 'http://example.com/entity/1',
          name: 'Test Entity 1',
          description: 'A test entity',
          entityType: 'http://pcdm.org/models#Collection',
          memberOf: null,
          rootCollection: null,
          metadataLicenseId: null,
          contentLicenseId: null,
          createdAt: new Date('2024-01-01'),
          updatedAt: new Date('2024-01-01'),
        },
        {
          rocrateId: 'http://example.com/entity/2',
          name: 'Test Entity 2',
          description: 'Another test entity',
          entityType: 'http://pcdm.org/models#Object',
          memberOf: null,
          rootCollection: null,
          metadataLicenseId: null,
          contentLicenseId: null,
          createdAt: new Date('2024-01-01'),
          updatedAt: new Date('2024-01-01'),
        },
      ];

      const mockSearchResponse = {
        body: {
          took: 10,
          hits: {
            total: { value: 2 },
            hits: [
              {
                _score: 1.5,
                _source: {
                  rocrateId: 'http://example.com/entity/1',
                },
                highlight: {
                  name: ['<em>Test</em> Entity 1'],
                },
              },
              {
                _score: 1.2,
                _source: {
                  rocrateId: 'http://example.com/entity/2',
                },
                highlight: {
                  description: ['Another <em>test</em> entity'],
                },
              },
            ],
          },
          aggregations: {
            entityType: {
              buckets: [
                { key: 'http://pcdm.org/models#Collection', doc_count: 1 },
                { key: 'http://pcdm.org/models#Object', doc_count: 1 },
              ],
            },
          },
        },
      };

      // @ts-expect-error TS is looking at the wrong function signature
      opensearch.search.mockResolvedValue(mockSearchResponse);
      // @ts-expect-error TS is looking at the wrong function signature
      prisma.entity.findMany.mockResolvedValue(mockEntities);

      const response = await fastify.inject({
        method: 'POST',
        url: '/search',
        payload: {
          query: 'test',
          searchType: 'basic',
        },
      });

      expect(response.statusCode).toBe(200);
      const body = JSON.parse(response.body);
      expect(body).toMatchSnapshot();

      // Verify database was queried with correct rocrateIds
      expect(prisma.entity.findMany).toHaveBeenCalledWith({
        where: {
          rocrateId: {
            in: ['http://example.com/entity/1', 'http://example.com/entity/2'],
          },
        },
      });

      expect(opensearch.search).toHaveBeenCalledWith({
        index: 'entities',
        body: {
          query: {
            bool: {
              must: [
                {
                  multi_match: {
                    query: 'test',
                    fields: ['name^2', 'description'],
                    type: 'best_fields',
                    fuzziness: 'AUTO',
                  },
                },
              ],
              filter: [],
            },
          },
          aggs: {
            inLanguage: { terms: { field: 'inLanguage', size: 20 } },
            mediaType: { terms: { field: 'mediaType', size: 20 } },
            communicationMode: { terms: { field: 'communicationMode', size: 20 } },
            entityType: { terms: { field: 'entityType', size: 20 } },
          },
          highlight: {
            fields: {
              name: {},
              description: {},
            },
          },
          sort: undefined,
          from: 0,
          size: 100,
        },
      });
    });

    it('should perform advanced search with query string', async () => {
      const mockSearchResponse = {
        body: {
          took: 5,
          hits: {
            total: { value: 0 },
            hits: [],
          },
          aggregations: {},
        },
      };

      // @ts-expect-error TS is looking at the wrong function signature
      opensearch.search.mockResolvedValue(mockSearchResponse);
      prisma.entity.findMany.mockResolvedValue([]);

      const response = await fastify.inject({
        method: 'POST',
        url: '/search',
        payload: {
          query: 'name:test AND description:entity',
          searchType: 'advanced',
        },
      });

      expect(response.statusCode).toBe(200);
      expect(opensearch.search).toHaveBeenCalledWith({
        index: 'entities',
        body: {
          query: {
            bool: {
              must: [
                {
                  query_string: {
                    query: 'name:test AND description:entity',
                    fields: ['name^2', 'description'],
                    default_operator: 'AND',
                  },
                },
              ],
              filter: [],
            },
          },
          aggs: {
            inLanguage: { terms: { field: 'inLanguage', size: 20 } },
            mediaType: { terms: { field: 'mediaType', size: 20 } },
            communicationMode: { terms: { field: 'communicationMode', size: 20 } },
            entityType: { terms: { field: 'entityType', size: 20 } },
          },
          highlight: {
            fields: {
              name: {},
              description: {},
            },
          },
          sort: undefined,
          from: 0,
          size: 100,
        },
      });
    });

    it('should apply filters correctly', async () => {
      const mockSearchResponse = {
        body: {
          took: 8,
          hits: {
            total: { value: 0 },
            hits: [],
          },
          aggregations: {},
        },
      };

      // @ts-expect-error TS is looking at the wrong function signature
      opensearch.search.mockResolvedValue(mockSearchResponse);
      prisma.entity.findMany.mockResolvedValue([]);

      const response = await fastify.inject({
        method: 'POST',
        url: '/search',
        payload: {
          query: 'test',
          filters: {
            entityType: ['http://pcdm.org/models#Collection'],
            inLanguage: ['en', 'fr'],
          },
        },
      });

      expect(response.statusCode).toBe(200);
      const expectedFilters = [
        {
          terms: {
            entityType: ['http://pcdm.org/models#Collection'],
          },
        },
        {
          terms: {
            inLanguage: ['en', 'fr'],
          },
        },
      ];

      expect(opensearch.search).toHaveBeenCalledWith(
        expect.objectContaining({
          body: expect.objectContaining({
            query: expect.objectContaining({
              bool: expect.objectContaining({
                filter: expectedFilters,
              }),
            }),
          }),
        }),
      );
    });

    it('should handle geospatial search with bounding box', async () => {
      const mockSearchResponse = {
        body: {
          took: 12,
          hits: {
            total: { value: 0 },
            hits: [],
          },
          aggregations: {
            geohash_grid: {
              buckets: [
                { key: 'gbsuv', doc_count: 3 },
                { key: 'gbsvb', doc_count: 1 },
              ],
            },
          },
        },
      };

      // @ts-expect-error TS is looking at the wrong function signature
      opensearch.search.mockResolvedValue(mockSearchResponse);
      prisma.entity.findMany.mockResolvedValue([]);

      const response = await fastify.inject({
        method: 'POST',
        url: '/search',
        payload: {
          query: 'test',
          boundingBox: {
            topRight: { lat: 51.5, lng: 0.1 },
            bottomLeft: { lat: 51.4, lng: 0.0 },
          },
          geohashPrecision: 5,
        },
      });

      expect(response.statusCode).toBe(200);
      const body = JSON.parse(response.body) as { geohashGrid: Record<string, number> };
      expect(body.geohashGrid).toEqual({
        gbsuv: 3,
        gbsvb: 1,
      });

      expect(opensearch.search).toHaveBeenCalledWith(
        expect.objectContaining({
          body: expect.objectContaining({
            query: expect.objectContaining({
              bool: expect.objectContaining({
                filter: [
                  {
                    geo_bounding_box: {
                      location: {
                        top_left: { lat: 51.5, lon: 0.0 },
                        bottom_right: { lat: 51.4, lon: 0.1 },
                      },
                    },
                  },
                ],
              }),
            }),
            aggs: expect.objectContaining({
              geohash_grid: {
                geohash_grid: {
                  field: 'location',
                  precision: 5,
                  bounds: {
                    top_left: { lat: 51.5, lon: 0.0 },
                    bottom_right: { lat: 51.4, lon: 0.1 },
                  },
                },
              },
            }),
          }),
        }),
      );
    });

    it('should handle pagination and sorting', async () => {
      const mockSearchResponse = {
        body: {
          took: 6,
          hits: {
            total: { value: 0 },
            hits: [],
          },
          aggregations: {},
        },
      };

      // @ts-expect-error TS is looking at the wrong function signature
      opensearch.search.mockResolvedValue(mockSearchResponse);
      prisma.entity.findMany.mockResolvedValue([]);

      const response = await fastify.inject({
        method: 'POST',
        url: '/search',
        payload: {
          query: 'test',
          limit: 50,
          offset: 20,
          sort: 'name',
          order: 'desc',
        },
      });

      expect(response.statusCode).toBe(200);
      expect(opensearch.search).toHaveBeenCalledWith(
        expect.objectContaining({
          body: expect.objectContaining({
            from: 20,
            size: 50,
            sort: [{ 'name.keyword': 'desc' }],
          }),
        }),
      );
    });

    it('should handle opensearch errors', async () => {
      opensearch.search.mockRejectedValue(new Error('OpenSearch connection failed'));

      const response = await fastify.inject({
        method: 'POST',
        url: '/search',
        payload: {
          query: 'test',
        },
      });

      expect(response.statusCode).toBe(500);
      const body = JSON.parse(response.body);
      expect(body).toMatchSnapshot();
    });

    it('should validate required query parameter', async () => {
      const response = await fastify.inject({
        method: 'POST',
        url: '/search',
        payload: {},
      });

      expect(response.statusCode).toBe(400);
    });

    it('should skip entities not found in database and log warning', async () => {
      const mockEntities = [
        {
          rocrateId: 'http://example.com/entity/1',
          name: 'Test Entity 1',
          description: 'A test entity',
          entityType: 'http://pcdm.org/models#Collection',
          memberOf: null,
          rootCollection: null,
          metadataLicenseId: null,
          contentLicenseId: null,
          createdAt: new Date('2024-01-01'),
          updatedAt: new Date('2024-01-01'),
        },
        // Entity 2 is missing from database
      ];

      const mockSearchResponse = {
        body: {
          took: 10,
          hits: {
            total: { value: 2 },
            hits: [
              {
                _score: 1.5,
                _source: {
                  rocrateId: 'http://example.com/entity/1',
                },
                highlight: {},
              },
              {
                _score: 1.2,
                _source: {
                  rocrateId: 'http://example.com/entity/2',
                },
                highlight: {},
              },
            ],
          },
          aggregations: {},
        },
      };

      // @ts-expect-error TS is looking at the wrong function signature
      opensearch.search.mockResolvedValue(mockSearchResponse);
      // @ts-expect-error TS is looking at the wrong function signature
      prisma.entity.findMany.mockResolvedValue(mockEntities);

      const response = await fastify.inject({
        method: 'POST',
        url: '/search',
        payload: {
          query: 'test',
        },
      });

      expect(response.statusCode).toBe(200);
      const body = JSON.parse(response.body);

      expect(body).toMatchSnapshot();
    });

    it('should handle missing rocrateId in search hit', async () => {
      const mockSearchResponse = {
        body: {
          took: 5,
          hits: {
            total: { value: 1 },
            hits: [
              {
                _score: 1.5,
                _source: {
                  // Missing rocrateId
                },
              },
            ],
          },
          aggregations: {},
        },
      };

      // @ts-expect-error TS is looking at the wrong function signature
      opensearch.search.mockResolvedValue(mockSearchResponse);
      prisma.entity.findMany.mockResolvedValue([]);

      const response = await fastify.inject({
        method: 'POST',
        url: '/search',
        payload: {
          query: 'test',
        },
      });

      expect(response.statusCode).toBe(500);
      console.log(response.body);
    });

    it('should handle rocrateId explicitly set to undefined', async () => {
      const mockSearchResponse = {
        body: {
          took: 5,
          hits: {
            total: { value: 1 },
            hits: [
              {
                _score: 1.5,
                _source: {
                  rocrateId: undefined,
                  name: 'Test Entity',
                },
              },
            ],
          },
          aggregations: {},
        },
      };

      // @ts-expect-error TS is looking at the wrong function signature
      opensearch.search.mockResolvedValue(mockSearchResponse);

      const response = await fastify.inject({
        method: 'POST',
        url: '/search',
        payload: {
          query: 'test',
        },
      });

      expect(response.statusCode).toBe(500);
      const body = JSON.parse(response.body);
      expect(body).toMatchSnapshot();
    });

    it('should handle invalid search response with missing hits data', async () => {
      const mockSearchResponse = {
        body: {
          took: 5,
          // Missing hits object
        },
      };

      // @ts-expect-error TS is looking at the wrong function signature
      opensearch.search.mockResolvedValue(mockSearchResponse);

      const response = await fastify.inject({
        method: 'POST',
        url: '/search',
        payload: {
          query: 'test',
        },
      });

      expect(response.statusCode).toBe(500);
      const body = JSON.parse(response.body) as { error: { code: string; message: string } };
      expect(body.error.code).toBe('INTERNAL_ERROR');
      expect(body.error.message).toBe('Search failed');
    });

    it('should apply custom entity transformers', async () => {
      // biome-ignore lint/suspicious/noExplicitAny: fine in tests
      const customTransformer = async (entity: any) => ({
        ...entity,
        tested: true,
      });

      await fastifyBefore();
      await fastify.register(searchRoute, {
        accessTransformer: AllPublicAccessTransformer,
        entityTransformers: [customTransformer],
      });

      const mockEntities = [
        {
          rocrateId: 'http://example.com/entity/1',
          name: 'Test Entity 1',
          description: 'A test entity',
          entityType: 'http://pcdm.org/models#Collection',
          memberOf: null,
          rootCollection: null,
          metadataLicenseId: null,
          contentLicenseId: null,
          createdAt: new Date('2024-01-01'),
          updatedAt: new Date('2024-01-01'),
        },
      ];

      const mockSearchResponse = {
        body: {
          took: 10,
          hits: {
            total: { value: 1 },
            hits: [
              {
                _score: 1.5,
                _source: {
                  rocrateId: 'http://example.com/entity/1',
                },
                highlight: {},
              },
            ],
          },
          aggregations: {},
        },
      };

      // @ts-expect-error TS is looking at the wrong function signature
      opensearch.search.mockResolvedValue(mockSearchResponse);
      // @ts-expect-error TS is looking at the wrong function signature
      prisma.entity.findMany.mockResolvedValue(mockEntities);

      const response = await fastify.inject({
        method: 'POST',
        url: '/search',
        payload: {
          query: 'test',
        },
      });

      const body = JSON.parse(response.body);

      expect(response.statusCode).toBe(200);
      expect(body).toMatchSnapshot();
    });

    it('should handle search response with no aggregations field', async () => {
      const mockEntities = [
        {
          rocrateId: 'http://example.com/entity/1',
          name: 'Test Entity 1',
          description: 'A test entity',
          entityType: 'http://pcdm.org/models#Collection',
          memberOf: null,
          rootCollection: null,
          metadataLicenseId: null,
          contentLicenseId: null,
          createdAt: new Date('2024-01-01'),
          updatedAt: new Date('2024-01-01'),
        },
      ];

      const mockSearchResponse = {
        body: {
          took: 10,
          hits: {
            total: { value: 1 },
            hits: [
              {
                _score: 1.5,
                _source: {
                  rocrateId: 'http://example.com/entity/1',
                },
                highlight: {},
              },
            ],
          },
          // No aggregations field
        },
      };

      // @ts-expect-error TS is looking at the wrong function signature
      opensearch.search.mockResolvedValue(mockSearchResponse);
      // @ts-expect-error TS is looking at the wrong function signature
      prisma.entity.findMany.mockResolvedValue(mockEntities);

      const response = await fastify.inject({
        method: 'POST',
        url: '/search',
        payload: {
          query: 'test',
        },
      });

      expect(response.statusCode).toBe(200);
      const body = JSON.parse(response.body);
      expect(body.facets).toBeUndefined();
      expect(body.geohashGrid).toBeUndefined();
      expect(body.entities).toHaveLength(1);
    });

    it('should handle search response with malformed aggregation buckets', async () => {
      const mockEntities = [
        {
          rocrateId: 'http://example.com/entity/1',
          name: 'Test Entity 1',
          description: 'A test entity',
          entityType: 'http://pcdm.org/models#Collection',
          memberOf: null,
          rootCollection: null,
          metadataLicenseId: null,
          contentLicenseId: null,
          createdAt: new Date('2024-01-01'),
          updatedAt: new Date('2024-01-01'),
        },
      ];

      const mockSearchResponse = {
        body: {
          took: 10,
          hits: {
            total: { value: 1 },
            hits: [
              {
                _score: 1.5,
                _source: {
                  rocrateId: 'http://example.com/entity/1',
                },
                highlight: {},
              },
            ],
          },
          aggregations: {
            entityType: {
              // buckets is not an array
              buckets: null,
            },
            inLanguage: {
              // buckets is undefined
            },
          },
        },
      };

      // @ts-expect-error TS is looking at the wrong function signature
      opensearch.search.mockResolvedValue(mockSearchResponse);
      // @ts-expect-error TS is looking at the wrong function signature
      prisma.entity.findMany.mockResolvedValue(mockEntities);

      const response = await fastify.inject({
        method: 'POST',
        url: '/search',
        payload: {
          query: 'test',
        },
      });

      expect(response.statusCode).toBe(200);
      const body = JSON.parse(response.body);
      expect(body.facets).toBeUndefined();
      expect(body.entities).toHaveLength(1);
    });

    it('should handle search response with malformed geohash aggregation buckets', async () => {
      const mockEntities = [
        {
          rocrateId: 'http://example.com/entity/1',
          name: 'Test Entity 1',
          description: 'A test entity',
          entityType: 'http://pcdm.org/models#Collection',
          memberOf: null,
          rootCollection: null,
          metadataLicenseId: null,
          contentLicenseId: null,
          createdAt: new Date('2024-01-01'),
          updatedAt: new Date('2024-01-01'),
        },
      ];

      const mockSearchResponse = {
        body: {
          took: 10,
          hits: {
            total: { value: 1 },
            hits: [
              {
                _score: 1.5,
                _source: {
                  rocrateId: 'http://example.com/entity/1',
                },
                highlight: {},
              },
            ],
          },
          aggregations: {
            geohash_grid: {
              // buckets is not an array
              buckets: 'invalid',
            },
          },
        },
      };

      // @ts-expect-error TS is looking at the wrong function signature
      opensearch.search.mockResolvedValue(mockSearchResponse);
      // @ts-expect-error TS is looking at the wrong function signature
      prisma.entity.findMany.mockResolvedValue(mockEntities);

      const response = await fastify.inject({
        method: 'POST',
        url: '/search',
        payload: {
          query: 'test',
        },
      });

      expect(response.statusCode).toBe(200);
      const body = JSON.parse(response.body);
      expect(body.geohashGrid).toBeUndefined();
      expect(body.entities).toHaveLength(1);
    });
  });
});
