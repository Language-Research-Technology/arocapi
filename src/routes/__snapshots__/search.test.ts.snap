// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Search Route > POST /search > should apply custom entity transformers 1`] = `
{
  "entities": [
    {
      "access": {
        "content": true,
        "metadata": true,
      },
      "contentLicenseId": null,
      "description": "A test entity",
      "entityType": "http://pcdm.org/models#Collection",
      "id": "http://example.com/entity/1",
      "memberOf": null,
      "metadataLicenseId": null,
      "name": "Test Entity 1",
      "rootCollection": null,
      "searchExtra": {
        "highlight": {},
        "score": 1.5,
      },
      "tested": true,
    },
  ],
  "searchTime": 10,
  "total": 1,
}
`;

exports[`Search Route > POST /search > should handle opensearch errors 1`] = `
{
  "error": {
    "code": "INTERNAL_ERROR",
    "message": "Search failed",
  },
}
`;

exports[`Search Route > POST /search > should handle rocrateId explicitly set to undefined 1`] = `
{
  "error": {
    "code": "INTERNAL_ERROR",
    "message": "Search failed",
  },
}
`;

exports[`Search Route > POST /search > should perform basic search successfully 1`] = `
{
  "entities": [
    {
      "access": {
        "content": true,
        "metadata": true,
      },
      "contentLicenseId": null,
      "description": "A test entity",
      "entityType": "http://pcdm.org/models#Collection",
      "id": "http://example.com/entity/1",
      "memberOf": null,
      "metadataLicenseId": null,
      "name": "Test Entity 1",
      "rootCollection": null,
      "searchExtra": {
        "highlight": {
          "name": [
            "<em>Test</em> Entity 1",
          ],
        },
        "score": 1.5,
      },
    },
    {
      "access": {
        "content": true,
        "metadata": true,
      },
      "contentLicenseId": null,
      "description": "Another test entity",
      "entityType": "http://pcdm.org/models#Object",
      "id": "http://example.com/entity/2",
      "memberOf": null,
      "metadataLicenseId": null,
      "name": "Test Entity 2",
      "rootCollection": null,
      "searchExtra": {
        "highlight": {
          "description": [
            "Another <em>test</em> entity",
          ],
        },
        "score": 1.2,
      },
    },
  ],
  "facets": {
    "entityType": [
      {
        "count": 1,
        "name": "http://pcdm.org/models#Collection",
      },
      {
        "count": 1,
        "name": "http://pcdm.org/models#Object",
      },
    ],
  },
  "searchTime": 10,
  "total": 2,
}
`;

exports[`Search Route > POST /search > should skip entities not found in database and log warning 1`] = `
{
  "entities": [
    {
      "access": {
        "content": true,
        "metadata": true,
      },
      "contentLicenseId": null,
      "description": "A test entity",
      "entityType": "http://pcdm.org/models#Collection",
      "id": "http://example.com/entity/1",
      "memberOf": null,
      "metadataLicenseId": null,
      "name": "Test Entity 1",
      "rootCollection": null,
      "searchExtra": {
        "highlight": {},
        "score": 1.5,
      },
    },
  ],
  "searchTime": 10,
  "total": 2,
}
`;
